File: ui/src/core/chains.ts
import type {BoardPost, PostLeg, ProofScheme} from '@/types'

/**
 * Which chains exist here, which ones a trade touches, and how a chain is named
 * to somebody reading a button.
 *
 * This file exists because the board's rule is per chain rather than per app:
 * you may act on an offer once you have PROVEN an address on every chain that
 * offer settles on. v1 had one pair, so the rule read as "both" — which is
 * exactly why it needed writing down as a list. With four pairs, "both" is wrong
 * in a way that is invisible: a ZTS↔ZTS offer would demand a Bitcoin proof for a
 * leg that does not exist, and a Solana one would demand nothing at all.
 *
 * So the requirement is DERIVED from the offer rather than asserted beside it.
 * `chainsForPost` reads the post and says which chains it settles on; every gate
 * in the UI and the refusal in Go are written over that answer.
 *
 * Nothing here is trusted with anything. It decides which button is disabled and
 * which sentence names what is missing — the enforcement lives in
 * `handleBoardPublish` and `handleBoardTake`, which refuse to sign a post or a
 * take carrying an address this browser has not proven.
 */

export type ChainId = 'btc' | 'znn' | 'sol'

export interface Chain {
  id: ChainId
  /** How a button says it: "Prove Bitcoin address". A chain, not a wallet —
   *  UniSat, Syrius and Phantom are how you prove one, not what is proven. */
  label: string
  /** What a person types amounts in. Not always the base unit: Solana is typed
   *  in SOL and held in lamports, which is the conversion Go owns. */
  unit: string
  /** True where a leg names a token as well as an amount. */
  tokenised: boolean
  /** The proof this chain's address is claimed with. One per chain: the scheme
   *  is a property of the signature format. */
  scheme: ProofScheme
  /** Which wallet this build drives for it, for a sentence that names one. */
  wallet: string
}

export const CHAINS: readonly Chain[] = [
  {id: 'btc', label: 'Bitcoin', unit: 'sat', tokenised: false, scheme: 'btc-ecdsa', wallet: 'UniSat'},
  {id: 'znn', label: 'Zenon', unit: '', tokenised: true, scheme: 'znn-ed25519', wallet: 'Syrius'},
  {id: 'sol', label: 'Solana', unit: 'SOL', tokenised: false, scheme: 'sol-ed25519', wallet: 'Phantom'},
]

export const CHAIN_IDS: readonly ChainId[] = CHAINS.map((c) => c.id)

export function chainFor(id: ChainId): Chain {
  const found = CHAINS.find((c) => c.id === id)
  if (!found) throw new Error(`${id} is not a chain this build knows`)
  return found
}

export function chainLabel(id: ChainId): string {
  return CHAINS.find((c) => c.id === id)?.label ?? id
}

export function schemeFor(id: ChainId): ProofScheme {
  return chainFor(id).scheme
}

/** The chain a proof scheme is about — the inverse of `schemeFor`. */
export function chainForScheme(scheme: ProofScheme): ChainId | '' {
  return CHAINS.find((c) => c.scheme === scheme)?.id ?? ''
}

/** "Bitcoin", "Bitcoin and Zenon", "Bitcoin, Zenon and Solana" — a list as
 *  somebody would say it, because these go into sentences. */
export function chainList(ids: readonly ChainId[]): string {
  const names = ids.map(chainLabel)
  if (names.length <= 1) return names[0] ?? ''
  return `${names.slice(0, -1).join(', ')} and ${names[names.length - 1]}`
}

/** Deduplicate and order a set of chains, so two sentences on the same page
 *  never name the same pair in a different order. */
export function orderChains(ids: readonly ChainId[]): ChainId[] {
  return CHAIN_IDS.filter((id) => ids.includes(id))
}

/** Which chains a trade settles on. A ZTS↔ZTS offer answers `['znn']`, which is
 *  correct: one proof covers both its legs. */
export function chainsForPost(post: BoardPost): ChainId[] {
  return orderChains([post.give?.chain, post.want?.chain].filter(Boolean) as ChainId[])
}

/** What is still unproven of what a trade needs. */
export function missingChains(need: readonly ChainId[], proven: readonly ChainId[]): ChainId[] {
  return CHAIN_IDS.filter((id) => need.includes(id) && !proven.includes(id))
}

/** A trade this build can post. Mirrors `Pairs` in wasm/chains.go, which is what
 *  actually refuses one; this list decides which buttons exist. */
export interface Pair {
  id: string
  label: string
  a: ChainId
  b: ChainId
  chains: ChainId[]
}

export const PAIRS: readonly Pair[] = [
  {id: 'btc-znn', label: 'BTC ⇄ ZTS', a: 'btc', b: 'znn', chains: ['btc', 'znn']},
  {id: 'znn-znn', label: 'ZTS ⇄ ZTS', a: 'znn', b: 'znn', chains: ['znn']},
  {id: 'sol-znn', label: 'SOL ⇄ ZTS', a: 'sol', b: 'znn', chains: ['sol', 'znn']},
  {id: 'sol-btc', label: 'SOL ⇄ BTC', a: 'sol', b: 'btc', chains: ['sol', 'btc']},
]

export function pairFor(id: string): Pair | undefined {
  return PAIRS.find((p) => p.id === id)
}

/** The stable name of a trade, in the order PAIRS declares it. */
export function pairId(a: ChainId, b: ChainId): string {
  const found = PAIRS.find((p) => (p.a === a && p.b === b) || (p.a === b && p.b === a))
  return found ? found.id : `${a}-${b}`
}

export function pairLabel(a: ChainId, b: ChainId): string {
  return pairFor(pairId(a, b))?.label ?? `${chainLabel(a)} ⇄ ${chainLabel(b)}`
}

/** The pairs somebody holding these proofs may post. Empty means the Post button
 *  has nothing to offer yet, which is a sentence rather than a dead control. */
export function postablePairs(proven: readonly ChainId[]): Pair[] {
  return PAIRS.filter((p) => p.chains.every((c) => proven.includes(c)))
}

/** The two well-known Zenon tokens, so a ZTS↔ZTS form has something to offer
 *  before anybody types a standard. Any other ZTS is typed in full. */
export const KNOWN_TOKENS: readonly {zts: string; name: string}[] = [
  {zts: 'zts1znnxxxxxxxxxxxxx9z4ulx', name: 'ZNN'},
  {zts: 'zts1qsrxxxxxxxxxxxxxmrhjll', name: 'QSR'},
]

/** How a ZTS reads in a row. Long standards are shortened rather than wrapped:
 *  a board of 26-character identifiers is a board nobody reads. */
export function tokenName(zts?: string): string {
  const t = (zts ?? '').trim() || KNOWN_TOKENS[0].zts
  const known = KNOWN_TOKENS.find((k) => k.zts === t)
  if (known) return known.name
  return t.length > 12 ? `${t.slice(0, 8)}…${t.slice(-4)}` : t
}

/** One leg of an offer, as a row says it: "10 ZNN", "400000 sat", "1.5 SOL". */
export function legLabel(leg?: PostLeg | {chain: ChainId; token?: string; amount: string}): string {
  if (!leg) return ''
  const amount = (leg.amount ?? '').trim()
  switch (leg.chain) {
    case 'btc':
      return `${amount} sat`
    case 'sol':
      return `${amount} SOL`
    case 'znn':
      return `${amount} ${tokenName(leg.token)}`
  }
  return amount
}

/** The unit a leg's amount is typed in, for a field's suffix. */
export function unitFor(chain: ChainId, token?: string): string {
  return chain === 'znn' ? tokenName(token) : chainFor(chain).unit
}

File: wasm/chains.go
package main

import (
	"fmt"
	"math/big"
	"strings"

	"github.com/zenon/ferry-web-v2/wasm/sol"
	"github.com/zenon/ferry-web-v2/wasm/znn"
)

// Which chains a leg can settle on, and what is true of each one regardless of
// which side of a trade it is.
//
// v1 of this app had one pair and so had no need for this file: "the Bitcoin
// leg" and "the Zenon leg" were two named fields, and every rule could be
// written against one or the other. Four pairs make that impossible — a
// ZTS↔ZTS swap has two Zenon legs and no Bitcoin one — so a leg now carries its
// chain and the rules are written against the pair.
//
// Nothing here decides anything about safety. It is the vocabulary: which
// chains exist, what their money is called, and which pairs this build will
// let somebody agree to.

// ChainID names a chain a leg can settle on.
type ChainID string

const (
	// ChainBTC is a P2SH hashed-timelock contract. Funded by an ordinary
	// payment; spent by an ephemeral key this app generates per swap.
	ChainBTC ChainID = "btc"
	// ChainZNN is go-zenon's `htlc` embedded contract, called with account
	// blocks the user's own wallet signs. Carries a ZTS, so two Zenon legs can
	// trade against each other as long as the tokens differ.
	ChainZNN ChainID = "znn"
	// ChainSOL is the solzen-htlc program: an escrow PDA holding lamports, with
	// both exits paying addresses fixed at creation.
	ChainSOL ChainID = "sol"
)

// Chain is what the rest of the code needs to know about a chain without
// caring which one it is.
type Chain struct {
	ID ChainID
	// Label is how a sentence names it: "Bitcoin", not "btc".
	Label string
	// Unit is what a person types amounts in — sat, ZNN-or-whatever-ZTS, SOL.
	// It is NOT always the base unit: Solana is typed in SOL and held in
	// lamports, which is exactly the conversion that has to happen in one place.
	Unit string
	// Decimals converts the typed unit into base units. Zero for Bitcoin,
	// because this app has always worked in satoshi. -1 for Zenon, whose
	// decimals belong to the token and are read from a node.
	Decimals int
	// Tokenised says a leg on this chain names a token as well as an amount.
	Tokenised bool
	// Proof is the scheme an address on this chain is claimed with on the board.
	Proof string
}

var chains = map[ChainID]Chain{
	ChainBTC: {ID: ChainBTC, Label: "Bitcoin", Unit: "sat", Decimals: 0, Proof: ProofBTC},
	ChainZNN: {ID: ChainZNN, Label: "Zenon", Unit: "", Decimals: -1, Tokenised: true, Proof: ProofZNN},
	ChainSOL: {ID: ChainSOL, Label: "Solana", Unit: "SOL", Decimals: 9, Proof: ProofSOL},
}

// ChainOrder is every chain, in the order sentences and lists name them. A
// fixed order so two sentences on the same page never disagree.
var ChainOrder = []ChainID{ChainBTC, ChainZNN, ChainSOL}

// KnownChain reports whether this build can settle a leg here.
func KnownChain(c ChainID) bool { _, ok := chains[c]; return ok }

// ChainOf returns a chain's properties, or an error naming what was asked for.
func ChainOf(c ChainID) (Chain, error) {
	got, ok := chains[c]
	if !ok {
		return Chain{}, fmt.Errorf("%q is not a chain this build knows: it is %s",
			c, strings.Join(chainIDStrings(), ", "))
	}
	return got, nil
}

// ChainLabel names a chain for a sentence, falling back to the raw id so a
// record from a newer build still renders something.
func ChainLabel(c ChainID) string {
	if got, ok := chains[c]; ok {
		return got.Label
	}
	return string(c)
}

func chainIDStrings() []string {
	out := make([]string, 0, len(ChainOrder))
	for _, c := range ChainOrder {
		out = append(out, string(c))
	}
	return out
}

// ---------- pairs ----------

// Pair is a trade this build will let two people agree to.
//
// A list rather than "any two chains", because two of the four combinations
// that would produce are not trades: BTC against BTC is a transfer with extra
// steps, and ZTS against the SAME ZTS is the same. The one that IS allowed and
// looks like it should not be — Zenon against Zenon — is allowed precisely
// because the tokens differ, which is checked separately (see Pair.Check).
type Pair struct {
	// A and B are the two chains, in the order the label names them. A trade
	// has no inherent direction: which side a given user is on is their leg's
	// Dir, not a property of the pair.
	A, B ChainID
	// Label is how the pair is offered: "BTC ⇄ ZNN".
	Label string
}

// Pairs is every trade this build supports.
var Pairs = []Pair{
	{A: ChainBTC, B: ChainZNN, Label: "BTC ⇄ ZTS"},
	{A: ChainZNN, B: ChainZNN, Label: "ZTS ⇄ ZTS"},
	{A: ChainSOL, B: ChainZNN, Label: "SOL ⇄ ZTS"},
	{A: ChainSOL, B: ChainBTC, Label: "SOL ⇄ BTC"},
}

// PairID is the stable name a pair travels under, in the order Pairs declares
// it — so "sol-btc" whichever side of it a given user is on. Derived from the
// chains rather than stored beside them, because a stored id is a second thing
// that can disagree with the legs it claims to summarise.
func PairID(a, b ChainID) string {
	for _, p := range Pairs {
		if (p.A == a && p.B == b) || (p.A == b && p.B == a) {
			return string(p.A) + "-" + string(p.B)
		}
	}
	return string(a) + "-" + string(b)
}

// SupportedPair reports whether these two chains may trade against each other.
func SupportedPair(a, b ChainID) bool {
	for _, p := range Pairs {
		if (p.A == a && p.B == b) || (p.A == b && p.B == a) {
			return true
		}
	}
	return false
}

// PairLabel names a pair for a button.
func PairLabel(a, b ChainID) string {
	for _, p := range Pairs {
		if (p.A == a && p.B == b) || (p.A == b && p.B == a) {
			return p.Label
		}
	}
	return ChainLabel(a) + " ⇄ " + ChainLabel(b)
}

// CheckPair refuses a combination that is not a trade.
//
// The two-Zenon case is the whole reason this takes tokens as well as chains. A
// swap is only atomic because two DIFFERENT things are locked against one
// secret; ZNN for ZNN through two HTLCs is an elaborate way to pay yourself, and
// worse, both halves verify perfectly, so nothing later would catch it.
func CheckPair(a, b ChainID, tokenA, tokenB string) error {
	if !KnownChain(a) {
		_, err := ChainOf(a)
		return err
	}
	if !KnownChain(b) {
		_, err := ChainOf(b)
		return err
	}
	if !SupportedPair(a, b) {
		return fmt.Errorf("this build does not swap %s against %s: it does %s",
			ChainLabel(a), ChainLabel(b), pairLabels())
	}
	if a == ChainZNN && b == ChainZNN {
		ta, tb := resolveToken(tokenA), resolveToken(tokenB)
		if ta == tb {
			return fmt.Errorf("both legs of this swap are %s, so there is nothing being "+
				"traded — a Zenon-to-Zenon swap needs two different tokens", ta)
		}
	}
	return nil
}

func pairLabels() string {
	out := make([]string, 0, len(Pairs))
	for _, p := range Pairs {
		out = append(out, p.Label)
	}
	return strings.Join(out, ", ")
}

// resolveToken applies the rule the creation form has always had: a blank token
// on a Zenon leg means ZNN.
func resolveToken(t string) string {
	t = strings.TrimSpace(t)
	if t == "" {
		return znn.ZnnTokenStandard
	}
	return t
}

// ---------- amounts ----------

// baseUnits converts a typed amount into a chain's base units.
//
// `decimals` is the chain's when it has fixed ones and the TOKEN's when the
// chain is Zenon, which is why it is a parameter rather than read from the map:
// a ZTS with 0 decimals and one with 8 are both ordinary, and getting it from
// anywhere but the issuing token is how a swap verifies against the wrong
// number.
func baseUnits(typed string, decimals int) (*big.Int, error) {
	return decimalToBaseUnits(strings.TrimSpace(typed), decimals)
}

// displayUnits is the inverse: base units back to the figure a person typed,
// trailing zeroes trimmed. Used only for display, never for a comparison.
func displayUnits(base *big.Int, decimals int) string {
	if decimals <= 0 {
		return base.String()
	}
	scale := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(decimals)), nil)
	whole, frac := new(big.Int).QuoRem(base, scale, new(big.Int))
	if frac.Sign() == 0 {
		return whole.String()
	}
	digits := strings.TrimRight(fmt.Sprintf("%0*s", decimals, frac.String()), "0")
	return whole.String() + "." + digits
}

// solLamports converts a typed SOL amount into lamports, refusing anything that
// will not fit — a uint64 of lamports is 18 billion SOL, so overflow means the
// figure is a mistake rather than a large trade.
func solLamports(typed string) (uint64, error) {
	n, err := baseUnits(typed, 9)
	if err != nil {
		return 0, fmt.Errorf("Solana amount: %w", err)
	}
	if n.Sign() <= 0 {
		return 0, fmt.Errorf("%q is not an amount of SOL", typed)
	}
	if !n.IsUint64() {
		return 0, fmt.Errorf("%q is more SOL than exists", typed)
	}
	return n.Uint64(), nil
}

// solDisplay renders lamports as SOL.
func solDisplay(lamports uint64) string {
	return displayUnits(new(big.Int).SetUint64(lamports), 9)
}

// LamportsPerSol is re-exported so callers do not import sol/ for one constant.
const LamportsPerSol = sol.LamportsPerSol
